query-planner-js/src/composedSchema/buildComposedSchema.ts

```diff
@@ -0,0 +1,240 @@
+
+    // We currently only allow join spec directives on object types.
+    if (!isObjectType(type)) continue;
+
+    assert(
+      type.astNode,
+      `GraphQL type "${type.name}" should contain AST nodes`,
+    );
+
+    const ownerDirectiveArgs = getArgumentValuesForDirective(
+      ownerDirective,
+      type.astNode,
+    );
+
+    const typeMetadata: FederationTypeMetadata = ownerDirectiveArgs
+      ? {
+          serviceName: graphMap[ownerDirectiveArgs?.['graph']].serviceName,
+          keys: new MultiMap(),
+          isValueType: false,
+        }
+      : {
+          isValueType: true,
+        };
+
+    type.extensions = {
+      ...type.extensions,
+      federation: typeMetadata,
+    };
+
+    const typeDirectivesArgs = getArgumentValuesForRepeatableDirective(
+      typeDirective,
+      type.astNode,
+    );
+
+    assert(
+      !(typeMetadata.isValueType && typeDirectivesArgs.length >= 1),
+      `GraphQL type "${type.name}" cannot have a @${typeDirective.name} \
+directive without an $${ownerDirective.name} directive`,
+    );
+
+    for (const typeDirectiveArgs of typeDirectivesArgs) {
+      const serviceName = graphMap[typeDirectiveArgs['graph']].serviceName;
+
+      const keyFields = parseFieldSet(typeDirectiveArgs['key']);
+
+      typeMetadata.keys?.add(serviceName, keyFields);
+    }
+
+    for (const fieldDef of Object.values(type.getFields())) {
+      assert(
+        fieldDef.astNode,
+        `Field "${type.name}.${fieldDef.name}" should contain AST nodes`,
+      );
+
+      const fieldDirectiveArgs = getArgumentValuesForDirective(
+        fieldDirective,
+        fieldDef.astNode,
+      );
+
+      if (!fieldDirectiveArgs) continue;
+
+      const fieldMetadata: FederationFieldMetadata = {
+        serviceName: graphMap[fieldDirectiveArgs?.['graph']]?.serviceName,
+      };
+
+      fieldDef.extensions = {
+        ...fieldDef.extensions,
+        federation: fieldMetadata,
+      };
+
+      if (fieldDirectiveArgs) {
+        const { requires, provides } = fieldDirectiveArgs;
+
+        if (requires) {
+          fieldMetadata.requires = parseFieldSet(requires);
+        }
+
+        if (provides) {
+          fieldMetadata.provides = parseFieldSet(provides);
+        }
+      }
+    }
+  }
+
+  // We filter out schema elements that should not be exported to get to the
+  // API schema.
+
+  const schemaConfig = schema.toConfig();
+
+  return new GraphQLSchema({
+    ...schemaConfig,
+    types: schemaConfig.types.filter(isExported),
+    directives: schemaConfig.directives.filter(isExported),
+  });
+
+  // TODO: Implement the IsExported algorithm from the Core Schema spec.
+  function isExported(element: NamedSchemaElement) {
+    return !(isAssociatedWithFeature(element, coreName) || isAssociatedWithFeature(element, joinName))
+  }
+
+  function isAssociatedWithFeature(
+    element: NamedSchemaElement,
+    featureName: string,
+  ) {
+    return (
+      element.name === `${featureName}` ||
```

Comment:
So this is a bit overzealous: the core spec says that only *directives* can use the unprefixed feature name. This seems pretty reasonable to me: prefixed things are "obviously" weird, and directives are probably only being defined rarely and by some of your most GraphQL-expert folks, but everybody is going to add fields and such all over the place, so we are explicitly *not* banning you from having a random field or enum value or whatever called `join`.

So technically your code here removes types named `core` or `join` unnecessarily. Hopefully people put capital letters on their types so hopefully this won't hurt anyone... but this function is declared as taking in any named schema element, not just types and directives, so maybe somebody will use it to filter field names later. Anyway my suggestion is to just be a tad more "spec compliant" and only do the exact matching for directives.

Also this seems like an unnecessary use of backquote interpolation